File: src/main/java/com/zepben/vertxutils/routing/Route.java
/*
 * Copyright 2020 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package com.zepben.vertxutils.routing;

import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.zepben.annotations.EverythingIsNonnullByDefault;
import com.zepben.vertxutils.routing.handlers.DecodeBodyHandler;
import com.zepben.vertxutils.routing.handlers.PathParamsHandler;
import com.zepben.vertxutils.routing.handlers.QueryParamsHandler;
import com.zepben.vertxutils.routing.handlers.params.*;
import io.vertx.core.Handler;
import io.vertx.core.buffer.Buffer;
import io.vertx.core.http.HttpMethod;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.BodyHandler;

import javax.annotation.Nullable;
import java.util.Arrays;
import java.util.List;
import java.util.function.BiConsumer;

@SuppressWarnings("WeakerAccess")
@EverythingIsNonnullByDefault
public class Route {

    @Nullable private final String path;
    private final boolean hasRegexPath;
    private final ImmutableSet<HttpMethod> methods;
    private final ImmutableList<RouteHandler> handlers;
    private final ImmutableList<Handler<RoutingContext>> failureHandlers;
    private final boolean isPublic;

    public static Builder builder() {
        return new Builder();
    }

    private Route(@Nullable String path,
                  boolean hasRegexPath,
                  ImmutableSet<HttpMethod> methods,
                  ImmutableList<RouteHandler> handlers,
                  ImmutableList<Handler<RoutingContext>> failureHandlers,
                  boolean isPublic) {
        this.path = path;
        this.hasRegexPath = hasRegexPath;
        this.methods = methods;
        this.handlers = handlers;
        this.failureHandlers = failureHandlers;
        this.isPublic = isPublic;
    }

    /**
     * The path of the route.
     *
     * @return The path of the route.
     */
    @Nullable
    public String path() {
        return path;
    }

    /**
     * Set to true if the path uses regular expressions. Defaults to false.
     *
     * @return true is the path uses regular expressions.
     */
    public boolean hasRegexPath() {
        return hasRegexPath;
    }

    /**
     * The HTTP method for the route.
     *
     * @return The HTTP method for the route.
     */
    public Iterable<HttpMethod> methods() {
        return methods;
    }

    /**
     * Return a list of handlers for this route.
     * <p>
     * Remember to always call {@link RoutingContext#next()} to chain to your next handler if you have more than one.
     *
     * @return A list of handlers for this route.
     */
    public List<RouteHandler> handlers() {
        return handlers;
    }

    /**
     * The failure handler for the route.
     *
     * @return The failure handler for the route.
     */
    public List<Handler<RoutingContext>> failureHandlers() {
        return failureHandlers;
    }

    /**
     * Indicates if the route should be documented as a public route.
     *
     * @return true if the route is a publicly documented route.
     */
    public boolean isPublic() {
        return isPublic;
    }

    @SuppressWarnings({"WeakerAccess", "UnusedReturnValue"})
    public static class Builder {
        @Nullable private String path = null;
        private boolean hasRegexPath = false;
        private final ImmutableSet.Builder<HttpMethod> methods = ImmutableSet.builder();
        @Nullable private PathParamsHandler pathParamsHandler = null;
        @Nullable private QueryParamsHandler queryParamsHandler = null;
        @Nullable private BodyHandler bodyHandler = null;
        @Nullable private DecodeBodyHandler decodeBodyHandler = null;
        private final ImmutableList.Builder<RouteHandler> handlers = ImmutableList.builder();
        private final ImmutableList.Builder<Handler<RoutingContext>> failureHandlers = ImmutableList.builder();
        private boolean isPublic = true;

        private Builder() {
        }

        public Builder path(String path) {
            if (path.isEmpty())
                throw new IllegalArgumentException("path must not be empty");

            if (path.indexOf('%') >= 0)
                throw new IllegalArgumentException("formatted path must not contain a '%'");

            this.path = path;
            return this;
        }

        public Builder path(String pathFormat, PathParamRule<?>... rules) {
            int count = 0;
            for (int index = pathFormat.indexOf('%'); index >= 0; index = pathFormat.indexOf('%', index + 1)) {
                ++count;
                if ((index == 0)
                    || (index >= pathFormat.length() - 1)
                    || (pathFormat.charAt(index - 1) != ':')
                    || (pathFormat.charAt(index + 1) != 's')) {
                    throw new IllegalArgumentException("invalid use of % in path format string");
                }
            }

            if (count < rules.length)
                throw new IllegalArgumentException("too many path params");
            else if (count > rules.length)
                throw new IllegalArgumentException("missing path params");

            path(String.format(pathFormat, Arrays.stream(rules).map(ParamRule::name).toArray()));
            pathParamsHandler = new PathParamsHandler(rules);
            return this;
        }

        public Builder hasRegexPath(boolean hasRegexPath) {
            this.hasRegexPath = hasRegexPath;
            return this;
        }

        public Builder method(HttpMethod method) {
            methods.add(method);
            return this;
        }

        public Builder methods(HttpMethod... methods) {
            for (HttpMethod method : methods)
                method(method);

            return this;
        }

        public Builder queryParams(QueryParamRule<?>... rules) {
            queryParamsHandler = new QueryParamsHandler(rules);
            return this;
        }

        public Builder bodySizeLimit(long size) {
            if (bodyHandler == null)
                bodyHandler(BodyHandler.create());

            bodyHandler.setBodyLimit(size);
            return this;
        }

        public Builder uploadsDirectory(String path) {
            if (bodyHandler == null)
                bodyHandler(BodyHandler.create());

            bodyHandler.setUploadsDirectory(path);
            return this;
        }

        public Builder decodeBody(RequestValueConverter<Buffer, ?> bodyConverter) {
            return decodeBody(bodyConverter, true);
        }

        public Builder decodeBody(RequestValueConverter<Buffer, ?> bodyConverter, boolean bodyRequired) {
            if (bodyHandler == null)
                bodyHandler(BodyHandler.create());

            decodeBodyHandler(new DecodeBodyHandler(new BodyRule<>(bodyConverter, bodyRequired)));
            return this;
        }

        public Builder bodyHandler(BodyHandler handler) {
            bodyHandler = handler;
            return this;
        }

        public Builder decodeBodyHandler(DecodeBodyHandler handler) {
            decodeBodyHandler = handler;
            return this;
        }

        public Builder addHandler(RouteHandler handler) {
            handlers.add(handler);
            return this;
        }

        public Builder addHandler(Handler<RoutingContext> handler) {
            return addHandler(new RouteHandler(handler, false));
        }

        /**
         * Registers a blocking handler.
         * This makes the handler equivalent to being registered with {@link io.vertx.ext.web.Route#blockingHandler(Handler, boolean)}.
         * on the {@link RouteRegister} however the boolean ordered flag is set by the argument given to the route register.
         *
         * @param blockingHandler The handler that contains blocking code.
         * @return This builder.
         */
        public final Builder addBlockingHandler(Handler<RoutingContext> blockingHandler) {
            return addHandler(new RouteHandler(blockingHandler, true, null));
        }

        /**
         * Registers a blocking handler.
         * This makes the handler equivalent to being registered with {@link io.vertx.ext.web.Route#blockingHandler(Handler, boolean)}
         * on the {@link RouteRegister}.
         *
         * @param blockingHandler The handler that contains blocking code.
         * @return This builder.
         */
        public final Builder addBlockingHandler(Handler<RoutingContext> blockingHandler, boolean ordered) {
            return addHandler(new RouteHandler(blockingHandler, true, ordered));
        }

        public final Builder addFailureHandler(Handler<RoutingContext> failureHandler) {
            this.failureHandlers.add(failureHandler);
            return this;
        }

        public final <T extends Throwable> Builder addFailureHandler(Class<T> throwableClass, BiConsumer<T, RoutingContext> handler) {
            failureHandlers.add(new ExceptionHandler<>(throwableClass, handler));
            return this;
        }

        public Builder isPublic(boolean isPublic) {
            this.isPublic = isPublic;
            return this;
        }

        public Route build() {
            if (path != null && !hasRegexPath && path.charAt(0) != '/')
                throw new IllegalStateException("path must start with a /");

            ImmutableList.Builder<RouteHandler> allHandlers = ImmutableList.builder();
            if (bodyHandler != null) {
                allHandlers.add(new RouteHandler(bodyHandler, false));
                if (decodeBodyHandler != null)
                    allHandlers.add(new RouteHandler(decodeBodyHandler, false));
            }

            if (pathParamsHandler != null)
                allHandlers.add(new RouteHandler(pathParamsHandler, false));

            if (queryParamsHandler != null)
                allHandlers.add(new RouteHandler(queryParamsHandler, false));

            allHandlers.addAll(handlers.build());

            return new Route(
                path,
                hasRegexPath,
                methods.build(),
                allHandlers.build(),
                failureHandlers.build(),
                isPublic);
        }
    }
}

File: src/main/java/com/zepben/vertxutils/routing/Route.kt
/*
 * Copyright 2026 Zeppelin Bend Pty Ltd
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.zepben.vertxutils.routing

import com.zepben.vertxutils.routing.handlers.DecodeBodyHandler
import com.zepben.vertxutils.routing.handlers.PathParamsHandler
import com.zepben.vertxutils.routing.handlers.QueryParamsHandler
import com.zepben.vertxutils.routing.handlers.params.BodyRule
import com.zepben.vertxutils.routing.handlers.params.PathParamRule
import com.zepben.vertxutils.routing.handlers.params.QueryParamRule
import com.zepben.vertxutils.routing.handlers.params.RequestValueConverter
import io.vertx.core.Handler
import io.vertx.core.http.HttpMethod
import io.vertx.ext.web.RequestBody
import io.vertx.ext.web.RoutingContext
import io.vertx.ext.web.handler.BodyHandler

/**
 * @property path The path of the route.
 * @property hasRegexPath Set to true if the path uses regular expressions. Defaults to false.
 * @property methods The HTTP method for the route.
 * @property handlers A list of handlers for this route. Remember to always call [RoutingContext.next] to chain to your next handler if you have more than one.
 * @property failureHandlers The failure handlers for the route.
 * @property isPublic Indicates if the route should be documented as a public route. True if the route is a publicly documented route.
 */
class Route private constructor(
    val path: String?,
    val hasRegexPath: Boolean,
    val methods: Set<HttpMethod>,
    val handlers: List<RouteHandler>,
    val failureHandlers: List<Handler<RoutingContext>>,
    val isPublic: Boolean,
) {

    class Builder internal constructor() {

        private var path: String? = null
        private var hasRegexPath = false
        private val methods = mutableSetOf<HttpMethod>()
        private var pathParamsHandler: PathParamsHandler? = null
        private var queryParamsHandler: QueryParamsHandler? = null
        private var bodyHandler: BodyHandler? = null
        private var decodeBodyHandler: DecodeBodyHandler? = null
        private val handlers = mutableListOf<RouteHandler>()
        private val failureHandlers = mutableListOf<Handler<RoutingContext>>()
        private var isPublic = true

        fun path(path: String): Builder = also { builder ->
            require(!path.isEmpty()) { "path must not be empty" }
            require(path.indexOf('%') < 0) { "formatted path must not contain a '%'" }

            builder.path = path
        }

        fun path(pathFormat: String, vararg rules: PathParamRule<*>): Builder = apply {
            var count = 0
            var index = pathFormat.indexOf('%')
            while (index >= 0) {
                ++count
                require((index != 0) && (index < (pathFormat.length - 1)) && (pathFormat[index - 1] == ':') && (pathFormat[index + 1] == 's')) {
                    "invalid use of % in path format string"
                }
                index = pathFormat.indexOf('%', index + 1)
            }

            require(count >= rules.size) { "too many path params" }
            require(count <= rules.size) { "missing path params" }

            path(String.format(pathFormat, *rules.map { it.name }.toTypedArray()))
            pathParamsHandler = PathParamsHandler(*rules)
        }

        fun hasRegexPath(hasRegexPath: Boolean): Builder = also { builder ->
            builder.hasRegexPath = hasRegexPath
        }

        fun method(method: HttpMethod): Builder = apply {
            methods.add(method)
        }

        fun methods(vararg methods: HttpMethod): Builder = apply {
            for (method in methods)
                method(method)
        }

        fun queryParams(vararg rules: QueryParamRule<*>): Builder = apply {
            queryParamsHandler = QueryParamsHandler(*rules)
        }

        fun bodySizeLimit(size: Long): Builder = apply {
            if (bodyHandler == null)
                bodyHandler(BodyHandler.create())

            bodyHandler!!.setBodyLimit(size)
        }

        fun uploadsDirectory(path: String): Builder = apply {
            if (bodyHandler == null)
                bodyHandler(BodyHandler.create())

            bodyHandler!!.setUploadsDirectory(path)
        }

        fun decodeBody(bodyConverter: RequestValueConverter<RequestBody, *>, bodyRequired: Boolean = true): Builder = apply {
            if (bodyHandler == null)
                bodyHandler(BodyHandler.create())

            decodeBodyHandler(DecodeBodyHandler(BodyRule(bodyConverter, bodyRequired)))
        }

        fun bodyHandler(handler: BodyHandler): Builder = apply {
            bodyHandler = handler
        }

        fun decodeBodyHandler(handler: DecodeBodyHandler): Builder = apply {
            decodeBodyHandler = handler
        }

        fun addHandler(handler: RouteHandler): Builder = apply {
            handlers.add(handler)
        }

        fun addHandler(handler: Handler<RoutingContext>): Builder =
            addHandler(RouteHandler(handler, false))

        /**
         * Registers a blocking handler.
         * This makes the handler equivalent to being registered with [io.vertx.ext.web.Route.blockingHandler].
         * on the [RouteRegister] however the boolean ordered flag is set by the argument given to the route register.
         * 
         * @param blockingHandler The handler that contains blocking code.
         * @return This builder.
         */
        fun addBlockingHandler(blockingHandler: Handler<RoutingContext>): Builder =
            addHandler(RouteHandler(blockingHandler, true, null))

        /**
         * Registers a blocking handler.
         * This makes the handler equivalent to being registered with [io.vertx.ext.web.Route.blockingHandler]
         * on the [RouteRegister].
         * 
         * @param blockingHandler The handler that contains blocking code.
         * @return This builder.
         */
        fun addBlockingHandler(blockingHandler: Handler<RoutingContext>, ordered: Boolean): Builder =
            addHandler(RouteHandler(blockingHandler, true, ordered))

        fun addFailureHandler(failureHandler: Handler<RoutingContext>): Builder = apply {
            failureHandlers.add(failureHandler)
        }

        fun <T : Throwable> addFailureHandler(throwableClass: Class<T>, handler: (T, RoutingContext?) -> Unit): Builder = apply {
            failureHandlers.add(ExceptionHandler(throwableClass, handler))
        }

        fun isPublic(isPublic: Boolean): Builder = also { builder ->
            builder.isPublic = isPublic
        }

        fun build(): Route {
            check((path == null) || hasRegexPath || (path!![0] == '/')) { "path must start with a /" }

            val allHandlers = listOfNotNull(
                bodyHandler,
                decodeBodyHandler.takeIf { bodyHandler != null },
                pathParamsHandler,
                queryParamsHandler,
            ).map { RouteHandler(it, false) } +
                handlers

            return Route(
                path,
                hasRegexPath,
                methods,
                allHandlers,
                failureHandlers,
                isPublic,
            )
        }
    }

    companion object {

        fun builder(): Builder = Builder()

    }

}
